tests/miral/wayland_extensions.cpp

```diff
@@ -138,12 +137,71 @@ struct ClientGlobalEnumerator
         new_global,
         global_remove
     };
-
-    std::shared_ptr<InterfaceNames> interfaces = std::make_shared<InterfaceNames>();
 };
 
 wl_registry_listener constexpr ClientGlobalEnumerator::registry_listener;
 
+/// Runs the client enumerator in a subprocess.
+///
+/// Mir will allow clients that originate from the compositor itself to access
+/// any protocols that they like. However, we would like to test the case
+/// where an external client asks the compositor for exstensions that it does
+/// not have access to. To do this, we need to fork this process, record the
+/// interfaces that come out of the child process, and return them to the
+/// test code via a pipe.
+///
+/// \param server
+/// \returns the list of interfaces that are available to the process
+std::vector<std::string> run_client_enumerator(mir::Server& server)
+{
+    std::vector<std::string> interfaces;
+
+    int pipefd[2];
+    if (pipe(pipefd) == -1)
+    {
+        std::cerr << "run_client_enumerator: pipe failed" << std::endl;
+        return {};
+    }
+
+    auto const pid = fork();
+    if (pid == 0)
+    {
+        close(pipefd[0]);
+        dup2(pipefd[1], STDOUT_FILENO);
+        close(pipefd[1]);
+
+        auto const wayland_display = server.wayland_display();
+        setenv("WAYLAND_DISPLAY", wayland_display.value().c_str(),  true);
+
+        ClientGlobalEnumerator enumerator;
+        enumerator.run();
+        _exit(0);
+    }
+
+    close(pipefd[1]);
+    FILE *stream = fdopen(pipefd[0], "r");
+    if (!stream)
+    {
+        std::cerr << "run_client_enumerator: fdopen failed" << std::endl;
+        return {};
+    }
+
+    char line[256];
+    while (fgets(line, sizeof(line), stream))
+    {
+        size_t const len = strlen(line);
+        if (len && line[len - 1] == '\n')
+            line[len - 1] = '\0';
+
+        interfaces.emplace_back(line);
```

Comment:
Can't we simply pass two iterators to the string constructor?

```suggestion
        interfaces.emplace_back(line, line + len-1);
```